### ISSUE TITLE
[BUG] Incorrect gradients with nested dynamic-range loops in reverse-mode AD

### ISSUE THREAD

**Issue by mar-yan24** [NVIDIA/warp#1293]
### Bug Description

I am currently playing around with the Auto Differentiation in Warp and I faced some issues when running reverse AD. It produces incorrect gradient values when two dynamic-range loops are nested in a kernel. The initial forward pass is correct, but `tape.backward()` yields a gradient 2x the correct value for the outer variable

### Reproduction/MRE
```
import warp as wp
import numpy as np

wp.init()


@wp.func
def _apply_joint(xpos: wp.vec3, xquat: wp.quat, angle: float, axis: wp.vec3):
    half = angle * 0.5
    qloc = wp.quat(axis[0] * wp.sin(half), axis[1] * wp.sin(half), axis[2] * wp.sin(half), wp.cos(half))
    xquat = xquat * qloc
    return xpos, xquat


@wp.kernel
def loop_kernel(
    branch_ids: wp.array(dtype=int), branch_starts: wp.array(dtype=int),
    parent: wp.array(dtype=int), jntnum: wp.array(dtype=int),
    jntadr: wp.array(dtype=int), jnt_axis: wp.array(dtype=wp.vec3),
    body_off: wp.array(dtype=wp.vec3), qpos: wp.array2d(dtype=float),
    xpos: wp.array2d(dtype=wp.vec3), xquat: wp.array2d(dtype=wp.quat),
):
    worldid, branchid = wp.tid()
    start = branch_starts[branchid]
    end = branch_starts[branchid + 1]
    q = qpos[worldid]
    for i in range(start, end):
        bid = branch_ids[i]
        pid = parent[bid]
        p = body_off[bid]
        r = wp.quat(0.0, 0.0, 0.0, 1.0)
        if pid >= 0:
            p = wp.quat_rotate(xquat[worldid, pid], p) + xpos[worldid, pid]
            r = xquat[worldid, pid]
        adr = jntadr[bid]
        for j in range(jntnum[bid]):
            p, r = _apply_joint(p, r, q[adr + j], jnt_axis[adr + j])
        xpos[worldid, bid] = p
        xquat[worldid, bid] = wp.normalize(r)


@wp.kernel
def unrolled_kernel(
    branch_ids: wp.array(dtype=int), branch_starts: wp.array(dtype=int),
    parent: wp.array(dtype=int), jntnum: wp.array(dtype=int),
    jntadr: wp.array(dtype=int), jnt_axis: wp.array(dtype=wp.vec3),
    body_off: wp.array(dtype=wp.vec3), qpos: wp.array2d(dtype=float),
    xpos: wp.array2d(dtype=wp.vec3), xquat: wp.array2d(dtype=wp.quat),
):
    worldid, branchid = wp.tid()
    start = branch_starts[branchid]
    end = branch_starts[branchid + 1]
    q = qpos[worldid]
    for i in range(start, end):
        bid = branch_ids[i]
        pid = parent[bid]
        p = body_off[bid]
        r = wp.quat(0.0, 0.0, 0.0, 1.0)
        if pid >= 0:
            p = wp.quat_rotate(xquat[worldid, pid], p) + xpos[worldid, pid]
            r = xquat[worldid, pid]
        adr = jntadr[bid]
        n = jntnum[bid]
        if n >= 1:
            p, r = _apply_joint(p, r, q[adr + 0], jnt_axis[adr + 0])
        if n >= 2:
            p, r = _apply_joint(p, r, q[adr + 1], jnt_axis[adr + 1])
        xpos[worldid, bid] = p
        xquat[worldid, bid] = wp.normalize(r)


def run(kernel):
    # 3-body hinge chain: world -> body1 (1 joint) -> body2 (1 joint)
    inputs = dict(
        branch_ids=wp.array([1, 2], dtype=int),
        branch_starts=wp.array([0, 2], dtype=int),
        parent=wp.array([-1, 0, 1], dtype=int),
        jntnum=wp.array([0, 1, 1], dtype=int),
        jntadr=wp.array([0, 0, 1], dtype=int),
        jnt_axis=wp.array([[0, 1, 0], [0, 1, 0]], dtype=wp.vec3),
        body_off=wp.array([[0, 0, 0], [0, 0, 0], [0, 0, -0.5]], dtype=wp.vec3),
        qpos=wp.array([[0.5, -0.3]], dtype=float, requires_grad=True),
    )
    xq_np = np.zeros((1, 3, 4), dtype=np.float32)
    xq_np[0, 0] = [0, 0, 0, 1]
    outputs = dict(
        xpos=wp.zeros((1, 3), dtype=wp.vec3, requires_grad=True),
        xquat=wp.array(xq_np, dtype=wp.quat, requires_grad=True),
    )
    loss = wp.zeros(1, dtype=float, requires_grad=True)

    tape = wp.Tape()
    with tape:
        wp.launch(kernel, dim=(1, 1), inputs=[*inputs.values()], outputs=[*outputs.values()])
        wp.launch(_sum, dim=3, inputs=[outputs["xpos"], loss])
    loss.grad.fill_(1.0)
    tape.backward()
    return inputs["qpos"].grad.numpy().flatten()


@wp.kernel
def _sum(xpos: wp.array2d(dtype=wp.vec3), loss: wp.array(dtype=float)):
    i = wp.tid()
    if i > 0 and i < 3:
        p = xpos[0, i]
        wp.atomic_add(loss, 0, p[0] + p[1] + p[2])

print(f"Nested loop grad: {run(loop_kernel)}")
print(f"Unrolled grad: {run(unrolled_kernel)}")
print("should match, the nested loop adjoint is wrong")
```

### Output
Output (Warp 1.12.0, CUDA 12.9, GTX 1070)

Nested loop grad: [-0.4251503  0.       ]
Unrolled grad:    [-0.19907852  0.        ]

There is a mismatch here.

### Expected vs Actual Behavior
Both kernels compute the identical forward pass. Their gradients should match. However, the nested `for j in range(jntnum[bid])` form produces a result of -0.425, this is a little more than 2x the correct value from the forward run. This probably is from the adjoint which is double counting a gradient contribution? 

### System Information

Warp 1.12.0
CUDA Toolkit 12.9, Driver 12.9
Python 3.13.5
Windows 10, GTX 1070 (sm_61)

**Comment by mar-yan24** [comment]
documented limitation
